sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,180 @@
+    ptrdiff_t mio_pads_register_offset =
+        PADCTRL_MIO_PADS0_REG_OFFSET + (i * PADCTRL_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(padctrl->base_addr, mio_pads_register_offset,
+                        PADCTRL_ATTR_DEFAULT);
+  }
+
+  return kDifPadctrlInitErrorNone;
+}
+
+dif_padctrl_registers_lock_error_t dif_padctrl_registers_lock(
+    dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlRegistersLockErrorInvalidPadctrl;
+  }
+
+  mmio_region_write32(padctrl->base_addr, PADCTRL_REGEN_REG_OFFSET, 0);
+
+  return kDifPadctrlRegistersLockErrorNone;
+}
+
+dif_padctrl_enable_attr_error_t dif_padctrl_enable_attr(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t attr, bool enable) {
+  if (padctrl == NULL) {
+    return kDifPadctrlEnableErrorInvalidPadctrl;
+  }
+
+  if (attr > kDifPadctrlAttrLast) {
+    return kDifPadctrlEnableErrorInvalidAttr;
+  }
+
+  ptrdiff_t pad_base_register_offset;
+  uint32_t pad_register_index;
+  uint32_t pad_attr_field_index;
+
+  // Do pad index range check and calculate which register and attribute field
+  // within it needs to be modified based on the kind of pad.
+  if (kind == kDifPadctrlPadMio) {
+    if (index > (PADCTRL_PARAM_NMIOPADS - 1)) {
+      return kDifPadctrlEnableErrorInvalidPad;
+    }
+
+    pad_base_register_offset = PADCTRL_MIO_PADS0_REG_OFFSET;
+    pad_register_index = index / PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+    pad_attr_field_index = index % PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+  } else {
+    if (index > (PADCTRL_PARAM_NDIOPADS - 1)) {
+      return kDifPadctrlEnableErrorInvalidPad;
+    }
+
+    pad_base_register_offset = PADCTRL_DIO_PADS_REG_OFFSET;
+    pad_register_index = index / PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG;
+    pad_attr_field_index = index % PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG;
+  }
+
+  // Determine which bit the attribute field that is being modified starts at
+  // and which bit within the register controls the attribute being modified.
+  uint32_t attr_field_start_bit = pad_attr_field_index * PADCTRL_PARAM_ATTRDW;
+  uint32_t attr_bit_index = attr_field_start_bit + attr;
+
+  // Calculate register offset for the register containing the attribute field
+  // that is being modified.
+  ptrdiff_t pad_register_offset =
+      pad_base_register_offset +
+      (PADCTRL_REGISTER_WIDTH_IN_BYTES * pad_register_index);
+
+  if (enable) {
+    // Extract the attribute bits from the given pad register and attribute
+    // field
+    uint32_t attr_field_mask = PADCTRL_ATTR_FIELD_MASK << attr_field_start_bit;
+
+    uint32_t attr_reg_value =
+        mmio_region_read32(padctrl->base_addr, pad_register_offset);
+
+    uint32_t attr_field_value =
+        (attr_reg_value & attr_field_mask) >> attr_field_start_bit;
+
+    // Update the read attribute field with the attribute to enable
+    attr_field_value |= (1 << attr);
```

Comment:
I think some/most of it can be done in terms of `bitfield_get_field32` in `bitfield` library and `mmio` library API.

 Hmmm we probably should add `mmio_region_nonatomic_get_field32` to mirror `mmio_region_nonatomic_set_field32` in `mmio.h`, this could be a subject of a follow-up PR. Or probably could be done as a separate commit in this PR. If you prefer to leave it for later, I can at some point in the future do it myself.